dts/arm/atmel/samd5x.dtsi

```diff
@@ -30,6 +30,13 @@
 				compatible = "arm,armv7m-mpu";
 				reg = <0xe000ed90 0x40>;
 			};
+
+			itm: itm@e0000000 {
+				compatible = "arm,armv7m-itm";
+				reg = <0xe0000000 0x1000>;
+
+				// reference frequency depends on GCLK_CM4_TRACE channel
```

Comment:
Why do we need this node? I mean, where it is used?
Could you pointing out?

Reply:
It's used for `swo-ref-frequency` in `subsys/logging/backends/log_backend_swo.c` which isn't currently used here as discussed.

But it also serves as a point for boards to hang off `pinctrl` properties, see for example https://github.com/zephyrproject-rtos/zephyr/blob/f29377a12cb5dd2ee845f21b8319c0c4ee2cca40/boards/nxp/mimxrt1060_evk/mimxrt1060_evk.dts#L293-L296.

`subsys/logging/backends/log_backend_swo.c` then can pick this up and configure the pin muxing accordingly.

Reply:
Ok, I got it. With you reference I found that it is already possible to use directly the CPU frequency as the base for the SWO, see https://github.com/zephyrproject-rtos/zephyr/blob/f29377a12cb5dd2ee845f21b8319c0c4ee2cca40/subsys/logging/backends/Kconfig.swo#L17

In this case, maybe we don't need to add the **itm** entry. Instead, we could set the GCLK_CM4_TRACE peripheral to use GCLKGEN[0], which is GCLK_MAIN.

Do you think this could work? Is it possible to you to test this idea?

Reply:
> In this case, maybe we don't need to add the itm entry.

Not for the clocking parts, as discussed below, but there's still the issue of providing a point to attach the `pinctrl` properties that actually configure the SWO pin to, so I think adding it still makes sense since this also is how all the other SoCs work.

> Instead, we could set the GCLK_CM4_TRACE peripheral to use GCLKGEN[0], which is GCLK_MAIN.
> Do you think this could work? Is it possible to you to test this idea?

Yeah this works and I have just tested it again on my board (by putting `PCHCTRL[47].reg = GCLK_PCHCTRL_GEN_GCLK0 | GCLK_PCHCTRL_CHEN;` into the clock setup in `soc/atmel/sam0/common/soc_samd5x.c`) - you'll get the frequency you expect.

This effectively is what I proposed in the PR cover letter. The remaining question is whether or not this clock should always be enabled even when SWO/the ITM is not being used. Technically we could check for `status = "okay"` on the `itm` node, but it seems that none of the other boards that use this node (like `mimxrt1060_evk.dts`) set this either, so I'm not sure what the most idiomatic approach would be here. We could also just enable the clock indiscriminately.

Reply:
Hi @florolf ,

Thank you for testing this.

I see 2 options here:

- Option 1 - Easy going

1) Add a rule to check if the ITM node exists and at same time check if CONFIG_LOG_BACKEND_SWO is enabled to apply the PCHCTRL[47].reg = GCLK_PCHCTRL_GEN_GCLK0 | GCLK_PCHCTRL_CHEN; at soc/atmel/sam0/common/soc_samd5x.c, as you mentioned.
2) Keep the ITM node with the pinctrl definitions but **do not** enabled it. The current version of the swo driver do not check if node is enabled or not allowing hacks. ?!
3) Add a note that node is exists only to provide the pinctrl functionality and should be keep as is.
4) add an ASSERT rule to fail at swo backend when there is no ITM node defined.

This should be enough to the system detect the ITM and apply the pinctrl rule. At same time, it will only enable the clock if this specific backend is enabled and node is present.

 - Option 2 - 100% generic and will work for all vendors

The problem with the SWO backend is that it uses Kconfig instead devicetree. The ideal is that backkend should be reworked to get all frequency values from devicetree. That way, the binding could have a phandle which can be selected by the user. That could be a reference for any clock defined for this board, including the cpu_freq. This is far better then a fixed frequency today. A divider could be defined at devicetree too. Then, all this could be easily customized at board, depending of each HW.

1) Rework the devicetree binding to have a phandle instead a fixed frequency. Possible the **clocks** property should be used. This will possibly deprecate the CONFIG_LOG_BACKEND_SWO_REF_FREQ_HZ config.
2) Add the divider or output frequency property at devicetree binding with a default. This possibly deprecate the LOG_BACKEND_SWO_FREQ_HZ.
3) Add a rule to check if ITM is enabled over devicetree and at same time CONFIG_LOG_BACKEND_SWO is enabled to apply the `PCHCTRL[47].reg = GCLK_PCHCTRL_GEN_GCLK0 | GCLK_PCHCTRL_CHEN;` at `soc/atmel/sam0/common/soc_samd5x.c`, as you already checked.
4) enable the ITM node at board.
5) add an ASSERT rule to fail at swo backend when there is no ITM node or it is disabled.

The option 2 then allow you to customize everything based on the real platform clocks. It could track the cpu_freq and apply correctly the pinctrl. Besides the ITM node be active, the system can detect and apply the configs only if the swo backend is enabled. At same time, it detects if devicetree configuration will be missing. The pinctrl will be optionally applied now because it will be only valid if node is enabled and there is an entry.